pkg/appregistry/builder.go

```diff
@@ -0,0 +1,222 @@
+package appregistry
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"database/sql"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"k8s.io/klog"
+
+	"github.com/operator-framework/operator-registry/pkg/apprclient"
+	"github.com/operator-framework/operator-registry/pkg/sqlite"
+)
+
+type RegistryImageBuilder interface {
+	Build() error
+}
+
+type ImageAppender interface {
+	Append(from, to, layer string) error
+}
+
+type ImageAppendFunc func(from, to, layer string) error
+
+func (f ImageAppendFunc) Append(from, to, layer string) error {
+	return f(from, to, layer)
+}
+
+type AppregistryImageBuilder struct {
+	Appender ImageAppender
+
+	// options
+	From, To            string
+	AuthToken           string
+	AppRegistryEndpoint string
+	AppRegistryOrg      string
+	DatabasePath        string
+	CacheDir            string
+
+	// derived
+	CleanOutput bool
+	ManifestDir string
+	DatabaseDir string
+}
+
+func NewAppregistryImageBuilder(config *AppregistryBuildOptions, options ...AppregistryBuildOption) (*AppregistryImageBuilder, error) {
```

Comment:
Would it be more flexible to have a return type of `RegistryImageBuilder` (an interface)?